### PR TITLE
Enable configuration binding generator in TodosApi app

### PR DESCRIPTION
**Pull request by layomia**
<details>
<summary>Generated code (click to view)</summary>

```cs
// <auto-generated/>
#nullable enable

/// <summary>Generated helper providing an AOT and linking compatible implementation for configuration binding.</summary>
internal static class GeneratedOptionsBuilderBinder
{
    /// <summary>Registers the dependency injection container to bind <typeparamref name="TOptions"/> against the <see cref="global::Microsoft.Extensions.Configuration.IConfiguration"/> obtained from the DI service provider.</summary>
    public static global::Microsoft.Extensions.Options.OptionsBuilder<TOptions> BindConfiguration<TOptions>(this global::Microsoft.Extensions.Options.OptionsBuilder<TOptions> optionsBuilder, string configSectionPath, global::System.Action<global::Microsoft.Extensions.Configuration.BinderOptions>? configureOptions = null) where TOptions : class
    {
        if (optionsBuilder is null)
        {
            throw new global::System.ArgumentNullException(nameof(optionsBuilder));
        }

        if (configSectionPath is null)
        {
            throw new global::System.ArgumentNullException(nameof(configSectionPath));
        }

        optionsBuilder.Configure<global::Microsoft.Extensions.Configuration.IConfiguration>((obj, configuration) =>
        {
            global::Microsoft.Extensions.Configuration.IConfiguration section = string.Equals(string.Empty, configSectionPath, global::System.StringComparison.OrdinalIgnoreCase) ? configuration : configuration.GetSection(configSectionPath);
            global::Microsoft.Extensions.Configuration.Binder.SourceGeneration.CoreBindingHelper.BindCoreUntyped(section, obj, typeof(TOptions), configureOptions);
        });

        global::Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton<global::Microsoft.Extensions.Options.IOptionsChangeTokenSource<TOptions>, global::Microsoft.Extensions.Options.ConfigurationChangeTokenSource<TOptions>>(optionsBuilder.Services);
        return optionsBuilder;
    }
}

namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
{
    using Microsoft.Extensions.Configuration;
    using Microsoft.Extensions.DependencyInjection;
    using System;
    using System.Collections.Generic;
    using System.Globalization;
    using TodosApi;

    /// <summary>Provide core binding logic.</summary>
    internal static class CoreBindingHelper
    {
        public static void BindCoreUntyped(this IConfiguration configuration, object obj, Type type, Action<BinderOptions>? configureOptions)
        {
            if (configuration is null)
            {
                throw new ArgumentNullException(nameof(configuration));
            }

            BinderOptions? binderOptions = GetBinderOptions(configureOptions);

            if (!HasValueOrChildren(configuration))
            {
                return;
            }

            if (type == typeof(AppSettings))
            {
                var temp = (AppSettings)obj;
                BindCore(configuration, ref temp, binderOptions);
                return;
            }

            throw new global::System.NotSupportedException($"Unable to bind to type '{type}': generator did not detect the type as input.");
        }

        public static void BindCore(IConfiguration configuration, ref AppSettings obj, BinderOptions? binderOptions)
        {
            if (obj is null)
            {
                throw new ArgumentNullException(nameof(obj));
            }

            List<string>? temp = null;
            foreach (IConfigurationSection section in configuration.GetChildren())
            {
                switch (section.Key)
                {
                    case "ConnectionString":
                        {
                            obj.ConnectionString = configuration["ConnectionString"]!;
                        }
                        break;
                    case "JwtSigningKey":
                        {
                            obj.JwtSigningKey = configuration["JwtSigningKey"]!;
                        }
                        break;
                    case "SuppressDbInitialization":
                        {
                            if (configuration["SuppressDbInitialization"] is string stringValue3)
                            {
                                obj.SuppressDbInitialization = ParseBool(stringValue3, () => section.Path)!;
                            }
                        }
                        break;
                    default:
                        {
                            if (binderOptions?.ErrorOnUnknownConfiguration == true)
                            {
                                (temp ??= new List<string>()).Add($"'{section.Key}'");
                            }
                        }
                        break;
                }
            }

            if (temp is not null)
            {
                throw new InvalidOperationException($"'ErrorOnUnknownConfiguration' was set on the provided BinderOptions, but the following properties were not found on the instance of {typeof(AppSettings)}: {string.Join(", ", temp)}");
            }
        }

        public static bool HasValueOrChildren(IConfiguration configuration)
        {
            if ((configuration as IConfigurationSection)?.Value is not null)
            {
                return true;
            }
            return HasChildren(configuration);
        }

        public static bool HasChildren(IConfiguration configuration)
        {
            foreach (IConfigurationSection section in configuration.GetChildren())
            {
                return true;
            }
            return false;
        }

        public static BinderOptions? GetBinderOptions(Action<BinderOptions>? configureOptions)
        {
            if (configureOptions is null)
            {
                return null;
            }
            BinderOptions binderOptions = new();
            configureOptions(binderOptions);
            if (binderOptions.BindNonPublicProperties)
            {
                throw new global::System.NotSupportedException($"The configuration binding source generator does not support 'BinderOptions.BindNonPublicProperties'.");
            }
            return binderOptions;
        }

        public static bool ParseBool(string stringValue, Func<string?> getPath)
        {
            try
            {
                return bool.Parse(stringValue);
            }
            catch (Exception exception)
            {
                throw new InvalidOperationException($"Failed to convert configuration value at '{getPath()}' to type '{typeof(bool)}'.", exception);
            }
        }
    }
}
```
</details>

The [binding extension classes](https://github.com/dotnet/runtime/tree/4bb41a7b85e29686a3882e777f0320a504cdd8bd/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src) in `Microsoft.Extensions.Options.ConfigurationExtensions` are now trimmed out from the app.

When https://github.com/dotnet/runtime/pull/87935 flows through, `ConfigurationBinder` itself will be trimmed out.

### CODE REVIEW

**Comment by layomia** [comment]
cc @DamianEdwards @eerhardt.

**Comment by eerhardt** [comment]
We should also remove this workaround:

https://github.com/aspnet/Benchmarks/blob/8c03b2411b78c691b00ac1610aac82a6d8849709/build/singlefile-scenarios.yml#L35-L36

**Comment by eerhardt** [comment]
Looks like we can't make this change until https://github.com/dotnet/runtime/issues/88112 is fixed.

File: src/BenchmarksApps/TodosApi/AppSettings.cs [hunk 1]
```diff
@@ -33,22 +33,10 @@ internal static class AppSettingsExtensions
...
-        // Can't use the configuration binding source generator due to bug where it emits non-compiling code right now
-        // https://github.com/dotnet/runtime/issues/83600
-        var optionsBuilder = services.Configure<AppSettings>(configurationRoot.GetSection(nameof(AppSettings)))
-            .AddOptions<AppSettings>();
-
-        if (!hostEnvironment.IsBuild())
-        {
-            services.AddSingleton<IValidateOptions<AppSettings>, AppSettingsValidator>();
-            optionsBuilder.ValidateOnStart();
-        }
```
Comment (eerhardt) [inline src/BenchmarksApps/TodosApi/AppSettings.cs:45]: @DamianEdwards - I assume we still want the `!IsBuild()` part here.

Comment (DamianEdwards) [reply src/BenchmarksApps/TodosApi/AppSettings.cs:45]: Yes we do

**Comment by eerhardt** [comment]
> @eerhardt I'm planning to fix the merge conflict; would that interrupt any work you're doing here?

I rebased on main yesterday and resolved the merge conflict. I think just the above comment needs to be resolved, then this change can be merged.

**Review by eerhardt — APPROVED** [review]

**Pull request merged**